crates/paimon/src/spec/tokens.rs

```diff
@@ -0,0 +1,1246 @@
+            Keyword::SMALLINT,
+            Keyword::INT,
+            Keyword::INTEGER,
+            Keyword::BIGINT,
+            Keyword::FLOAT,
+            Keyword::DOUBLE,
+            Keyword::PRECISION,
+            Keyword::DATE,
+            Keyword::TIME,
+            Keyword::WITH,
+            Keyword::WITHOUT,
+            Keyword::LOCAL,
+            Keyword::ZONE,
+            Keyword::TIMESTAMP,
+            Keyword::TimestampLtz,
+            Keyword::INTERVAL,
+            Keyword::YEAR,
+            Keyword::MONTH,
+            Keyword::DAY,
+            Keyword::HOUR,
+            Keyword::MINUTE,
+            Keyword::SECOND,
+            Keyword::TO,
+            Keyword::ARRAY,
+            Keyword::MULTISET,
+            Keyword::MAP,
+            Keyword::ROW,
+            Keyword::NULL,
+            Keyword::RAW,
+            Keyword::LEGACY,
+            Keyword::NOT,
+        ]
+        .into_iter()
+        .collect()
+    }
+
+    pub fn is_keyword(s: &str) -> bool {
+        Keyword::variants().iter().any(|v| v.to_string() == s)
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub enum TokenType {
+    /// e.g. "ROW<"
+    BeginSubtype,
+    /// e.g. "ROW<..>"
+    EndSubtype,
+    /// e.g. "CHAR("
+    BeginParameter,
+    /// e.g. "CHAR(...)"
+    EndParameter,
+    /// e.g. "ROW<INT,"
+    ListSeparator,
+    /// e.g. "ROW<name INT 'Comment'"
+    LiteralString,
+    /// CHAR(12
+    LiteralInt,
+    /// e.g. "CHAR" or "TO"
+    Keyword,
+    /// e.g. "ROW<name" or "myCatalog.myDatabase"
+    Identifier,
+    /// e.g. "myCatalog.myDatabase."
+    IdentifierSeparator,
+}
+
+#[derive(Debug, PartialEq)]
+pub struct Token {
+    token_type: TokenType,
+    cursor_position: isize,
+    value: String,
+}
+
+impl Token {
+    pub fn new(token_type: TokenType, cursor_position: isize, value: String) -> Self {
+        Token {
+            token_type,
+            cursor_position,
+            value,
+        }
+    }
+}
+
+pub struct TokenParser {
+    input_string: String,
+    tokens: Vec<Token>,
+    last_valid_token: isize,
+    current_token: isize,
+}
+
+impl TokenParser {
+    pub fn new(input_string: String, tokens: Vec<Token>) -> Self {
+        TokenParser {
+            input_string,
+            tokens,
+            last_valid_token: -1,
+            current_token: -1,
+        }
+    }
+
+    pub fn parse_tokens(&mut self) -> Result<DataType, Error> {
+        let type_data = self.parse_type_with_nullability();
+        if self.has_remaining_tokens() {
+            self.next_token();
+            self.parsing_error(format!("Unexpected token: {}", self.token().value));
+        }
+        Ok(type_data)
+    }
+
+    pub fn last_cursor(&self) -> usize {
+        if self.last_valid_token < 0 {
+            0
+        } else {
+            self.tokens[self.last_valid_token as usize].cursor_position as usize + 1
+        }
+    }
+
+    pub fn parsing_error(&self, cause: String) -> ! {
+        panic!(
```

Comment:
Using panic directly here might not be ideal, but we can leave a TODO to explore a better approach in the future.

Reply:
Any TODO comment？